feat(ApplicationLauncher): Add Application Launcher Component

gilad215 opened the pull request:
fix #184

<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->
**What**:
Add an Application Launcher Component.

Managed to show all the launcher types in one storybook: `Grid/List Icons/No Icons` with the help of the knobs
<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
**Link to Storybook**:
[Storybook](https://rawgit.com/gilad215/patternfly-react/storybook/feature/applauncher/index.html?knob-Launcher%20Type=true&knob-Icons=true&knob-Label=Well%20done%21%20You%20successfully%20read%20this%20important%20alert%20message.&selectedKind=patternfly-react%2FApplication%20Framework%2FLauncher&selectedStory=Stateful%20Launcher&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:
- Figuring out the Icons names are a real pain, any easy way to do it?
e.g. - `applauncher-pf-link-icon pficon pficon-domain` - What is the name prop that i need to give to `<Icon />` in order to render this icon?

<!-- feel free to add additional comments -->

**TODO:**
- [x] Create Tests

sharvit commented on src/components/ApplicationLauncher/ApplicationLauncherItem.js:
```diff
@@ -0,0 +1,50 @@
+    >
+      <li className={classes} role="menuitem">
+        <a className="applauncher-pf-link" href="#" onClick={e => onClick(e)}>
+          <i className={icon} aria-hidden="true" />
```
```js
<Icon name="home" className="applauncher-pf-link-icon" />
```

gilad215 commented:
- Added Tests
- Added a Wrapper component and story

jeff-phillips-18 commented on src/components/ApplicationLauncher/ApplicationLauncherToggle.js:
```diff
@@ -0,0 +1,38 @@
+import { Button } from '../Button';
+
+const ApplicationLauncherToggle = ({ onClick, tooltipPlacement }) => {
+  const tooltip = <Tooltip id="tooltip">Application Launcher</Tooltip>;
```
`Application Launcher` should be replaced with a passed in prop.

jeff-phillips-18 commented on src/components/ApplicationLauncher/ApplicationLauncher.js:
```diff
@@ -0,0 +1,61 @@
+
+    this.state = {
+      showLauncher: false,
+      type: props.type
```
why add type to state rather than simply using the prop? It should only be mutable by the caller

jeff-phillips-18 commented on src/components/ApplicationLauncher/ApplicationLauncher.js:
```diff
@@ -0,0 +1,61 @@
+      {
+        'applauncher-pf dropdown dropdown-kebab-pf': this.state.type === 'list'
+      },
+      { open: this.state.showLauncher }
```
Would be better to always specify `applauncher-pf dropdown dropdown-kebab-pf` and add `applauncher-pf-block-list` if type === grid.

jeff-phillips-18 commented on src/components/ApplicationLauncher/ApplicationLauncherItem.js:
```diff
@@ -0,0 +1,60 @@
+  /** Icon Type */
+  icon: PropTypes.string.isRequired,
+  /** App Tooltip */
+  tooltip: PropTypes.string.isRequired,
```
Should this be required? I'm not sure all items will need a tooltip.

jeff-phillips-18 commented on src/components/ApplicationLauncher/Wrappers/ApplicationLauncherWrapper.js:
```diff
@@ -0,0 +1,45 @@
+};
+ApplicationLauncherWrapper.propTypes = {
+  /** Array of App Objects */
+  apps: PropTypes.array.isRequired,
```
Could we define a shape for this so callers no what fields are required?

gilad215 commented:
- Applied @jeff-phillips-18 suggestions

jeff-phillips-18 commented on packages/core/src/components/ApplicationLauncher/Wrappers/ApplicationLauncherWrapper.js:
```diff
@@ -0,0 +1,56 @@
+  /** Grid instead of List (List is Default) */
+  grid: PropTypes.bool,
+  /** Tooltip Placement */
+  tooltipPlacement: PropTypes.string
```
Should take a tooltip to pass to the ApplicationLauncher to pass to the ApplicationLauncherToggle

gilad215 commented:
- Fixed Toggle tooltip and added a Shape proptype for the apps array inside the wrapper.

jgiardino commented:
@patternfly/patternfly-react-ux 
Can someone can do a UX review of this component?

lizsurette commented:
@jgiardino I'll review this afternoon.

gilad215 commented:
- Added a Stateful Wrapper to handle the Open/Close and added a ClickedOutSide functionality
- Toggle Tooltip is now optional

@lizsurette Thanks for the review Liz
1. Since the size of the dropdowns are the same on the different nav bars, I believe the app launcher is supported on both. I see the same classnames of the toggle component for the short and tall masthead.
2. Tooltips are in the design section, If youre talking about the toggle tooltip, its now optional.
<img width="329" alt="screen shot 2018-05-23 at 12 18 06" src="https://user-images.githubusercontent.com/24938324/40415408-7c1ec8da-5e83-11e8-80de-b250fb810f17.png">

3. I added a Stateful wrapper and added a ClickedOutSide functionality to it, but its the responsibility of the consumer to handle these kinds of actions.

4. Removed the Arrow.

jeff-phillips-18 commented:
@lizsurette Can you open an issue against the design documentation to remove the arrow pointing to the icon when opened from the images included there?

lizsurette commented:
@gilad215 Thanks for the updates! This all looks great to me. I guess the one open item is just around the arrow that you so quickly and nicely removed :) @jeff-phillips-18 In looking at the arrow discrepancy closer, while it's not in the base PF on the Design tab it is in the Angular implementation so now I'm questioning if it's a bug in the base PF implementation! Thoughts?

jgiardino commented:
Sounds like a question for @mcarrano ;-)

The [design documentation](http://www.patternfly.org/pattern-library/application-framework/launcher/#design) also includes the arrow, but I think the [masthead design](http://www.patternfly.org/pattern-library/application-framework/masthead/#design) was more recently updated, and it shows the app launcher without the arrow.

mcarrano commented:
Yep. The design documentation for this pattern is out of date but the core implementation examples are correct. The [Masthead pattern](http://www.patternfly.org/pattern-library/application-framework/masthead/) has the correct design.

I will file an issue to update the Application Launcher mocks.

jeff-phillips-18 commented:
An issue needs to be filed against angular-patternfly as well...

lizsurette commented:
@mcarrano Thanks for confirming the correct design, Matt!

@gilad215 @jeff-phillips-18 Yep, this is good to go now from a UX point of view :)

@jeff-phillips-18 I logged the issue in angular-patternfly: https://github.com/patternfly/angular-patternfly/issues/741

jgiardino commented:
Great work on this, @gilad215! I have a few questions about it, and then some requests listed below…

1. What’s the difference between the two storybooks? 
2. In the [pf core example](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/application-launcher-nav-vertical.html#0), at 768px, the elements in the masthead are hidden and become available in the vertical navigation menu. Do you know if that behavior is possible with this component? Or should we add a separate issue to capture that?

**Requested updates:** 

1. Please provide this link in the storybook for the design documentation:
http://www.patternfly.org/pattern-library/application-framework/launcher/#design

2. The element `.applauncher-pf-block-list` is missing the class `.dropdown`, and the `<button>` is missing `.dropdown-toggle`.  These classes are used in a few different selectors that apply padding to the button. 

3. The button element should include an `aria-expanded` attribute, set to “true” when the menu is visible and “false” when the menu is hidden. 

4. Inside the button, the current pf core implementation includes several span elements that are missing. These are all the contents of the button in the pf core example:
    ```
    <span class="fa fa-th applauncher-pf-icon" aria-hidden="true"></span>
    <span class="dropdown-title">
      <span class="applauncher-pf-title">
        Application Launcher
        <span class="caret" aria-hidden="true"></span>
      </span>
    </span>
    ```
    In this implementation, there's only this one span:
    ```
    <span aria-describedby="tooltip" aria-hidden="true" class="fa fa- fa fa-th applauncher-pf-icon"></span>
    ```
    - Please update the class names so that they match what’s included for core
    - Please move the `aria-describedby` attribute to the parent `<a>`. This attribute has [better support for interactive elements](https://developer.paciellogroup.com/blog/2017/07/short-note-on-aria-label-aria-labelledby-and-aria-describedby/), like `<a>`.

5. The roles currently applied to the menu in the pf core example are actually incorrect. The way they’re applied for the [pf-react dropdown button](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Style=default&selectedKind=patternfly-react%2FWidgets%2FButton&selectedStory=DropdownButton&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel) are correct. 
    - `<li>` should have `role=“presentation”` 
    - `<a>` should have `role=“menuitem”`
    - [explanation about these roles](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menu-button/menu-button-links.html)

6. When I hover over menu items, the `<li>` gets the attribute `aria-describedby=“tooltip”`, but it should be applied to the `<a>` instead.  The value should also map to the `id` value of another element, but currently all menu items are referencing the same tooltip `id`. Could you update the example to show different tooltip text being used? It would be good to verify that this works as expected, but also good to provide an example that's closer to what consumers would be implementing.

gilad215 commented:
@jgiardino 
> In the pf core example, at 768px, the elements in the masthead are hidden and become available in the vertical navigation menu. Do you know if that behavior is possible with this component? Or should we add a separate issue to capture that?

Should be possible, in your example the launcher should not be rendered at a certain width, its the responsibility of the consumer to handle these kind of behaviours. 

1. Fixed doc link
2. Added `.dropdown-toggle` for the button, This div already has the `.dropdown` class for me: 
<img width="343" alt="screen shot 2018-05-27 at 10 58 38" src="https://user-images.githubusercontent.com/24938324/40584446-67c7a824-61a9-11e8-80e2-d88e1f7878b6.png">

3. Done
4. Done, but i'm not sure why we need those extra spans, take a look.
5. Done
6. I removed the id for the tooltip itself, theres an id for each of the Apps.

jgiardino commented:
Hi @gilad215,

Thanks for making updates requested above! There's just one element that still doesn't match what's included in pf core:

4. For this element: `<span aria-describedby="tooltip" aria-hidden="true" class="fa fa- fa fa-th applauncher-pf-icon"></span>`
    - It still includes a weird set of class names: `fa fa- fa fa-th applauncher-pf-icon` should just be `fa fa-th applauncher-pf-icon`
    - It still includes `aria-describedby="tooltip"`. It looks like the tooltip for this icon was removed? If that's the case, this attribute that references it should also be removed. If you do keep this attribute, it should be moved to the parent `<button>` element.

gilad215 commented:
@jgiardino fixed the element :)

dabeng commented:
Hi @gilad215 I find that launcher button will disappear when the width of window is less than 768px.

![i988yxd300](https://user-images.githubusercontent.com/5689075/40897606-9bd864a0-67ee-11e8-8686-ee1e0b928b61.gif)

jgiardino commented on packages/core/src/components/ApplicationLauncher/ApplicationLauncherToggle.js:
```diff
@@ -0,0 +1,65 @@
+          onClick={onClick}
+          bsStyle="link"
+          className="nav-item-iconic dropdown-toggle"
+          aria-describedby="tooltip"
```
Where is this value coming from? This should the the same value that displays as the tooltip `id` value, but it isn't obvious to me where the tooltip `id` value is being defined. When I look at the [react-bootstrap documentation](https://react-bootstrap.github.io/components/tooltips/), they show examples of tooltip like this `<Tooltip id={id}>` with `id` as a required prop.

gilad215 replied:
yeah, ill just change it:
`        overlay={<Tooltip id={tooltip}>{tooltip}</Tooltip>}
`
in the `OverlayTrigger`

jgiardino commented:
@gilad215 I had to build your branch locally to see the updates, and they match what I was expecting. Thanks for the updates! I did have a question about about the tooltip, that I included in an inline comment. 

Do you happen to know why the storybook isn't showing the latest updates? (adding @priley86 and @jeff-phillips-18 in case this is an issue with something in the repo, and also for feedback on my next comment...)

Re: @dabeng's comment:

> I find that launcher button will disappear when the width of window is less than 768px.

This behavior comes from pf-core css—the css from core will not show the app launcher icon at a width less than 768px.

But in pf-core, the menu items in the app launcher menu will also get added to the vertical nav when the window is less than 768px in width. I'm assuming that this component is intended to support only the toggle and menu that display when the window is greater than 768px. But my question for everyone (@gilad215, @priley86, @jeff-phillips-18, @mturley) is how should the menu items be added to the vertical nav? Would that functionality be added to the vertical nav component? Is that something the consumer would manage? Do we need another issue to capture this aspect of the app launcher behavior?

jeff-phillips-18 commented:
The items in right of the masthead (app launcher, help, user menu, notification drawer trigger) are meant to be added to the vertical nav at less than 768px. This is described in http://www.patternfly.org/pattern-library/navigation/vertical-navigation/#design in the Responsive States section.

gilad215 commented:
@jgiardino changed `overlay={<Tooltip id={tooltip}>{tooltip}</Tooltip>}` in the `OverlayTrigger`
Also updated the Storybook

jgiardino commented:
Awesome! Thanks, @gilad215! No further comments from me. Great job on this!

![bitmoji](https://render.bitstrips.com/v2/cpanel/95caaa1e-9572-4e53-8013-e3e6215ae489-fe7ba5ad-9c01-4dac-b094-998c3f204a3b-v1.png?transparent=1&palette=1&width=246)

We also have issue #382 to address getting the app launcher into the vertical nav at smaller window sizes.
